Restructure the main part of create.adoc (#670)
neo4j/docs-cypher: modules/ROOT/pages/clauses/create.adoc

[[query-create]]
= CREATE

== Introduction

The `CREATE` clause allows you to create nodes and relationships.
Similar to xref::clauses/match.adoc[`MATCH`], this is done using xref::patterns/index.adoc[patterns].

> However, the patterns you can use in `CREATE` are restricted to what is relevant to describe the newly created nodes and relationships.

Reviewer: I feel uneasy about this part.
You create elements by spelling out explicit nodes or paths: `(:A)`, `(:A)-[r:R]->({foo: "bar"})`.
You match on elements using nodes or path patterns. A path pattern represents the set of paths in the graph matching the pattern at the time of execution of the query.
The node `(:A)` is very different indeed from the node pattern `(:A)`, the lattern will also return node `(:A:B)` for example.
The syntax for a path happens to be valid syntax for a path pattern, but I'm not sure it is useful to present the path syntax as a restricted version of the path pattern syntax.
The same way the string "hello" is a valid regex that will match on the string "hello". When you ask the user for a string, you don't tell them to enter a restricted regular expression.
I know that we haven't made that distinction in our parser or in our AST[^1], but I think we should make it clearer to our users.

[^1]: I think we should differentiate it there as well but it's a different conversation.

[[create-nodes]]
== Syntax for nodes

The `CREATE` clause allows you to create one or more nodes.

> For each, you can specify labels and properties.

Reviewer 2: ```suggestion
Each node can be assigned labels and properties.
```

> You can give the nodes names that you can refer to later in the query.

Reviewer 2: When we say node "names" - we just mean labels, right? If so, I think it would be better to just stick with labels, as right now the 2 could be interpreted to mean to different things.

Author: I mean identifiers/variables, as in this case `charlie` and `oliver`. Maybe I should call it "identifier" or "variable" then.

Reviewer 2: I think "variable" is preferable.

Author: I added:

> You can bind each node to a variable that you can refer to later in the query.

xref:clauses/where.adoc[`WHERE`-subclauses] and xref:clauses/match.adoc#label-expression-match-or-expression[label expression] are not allowed.

> Instead, multiple labels are separated by colons.

Reviewer: This follows from my initial comment, `WHERE` only makes sense in a node pattern, not in a node.

.Query
[source, cypher, indent=0]
----
CREATE (charlie:Person:Actor {name: 'Charlie Sheen'}), (oliver:Person:Director {name: 'Oliver Stone'})
----

> Creates two nodes, each with a `name` and a label `Person`.

Reviewer 2: ```suggestion
Creates two nodes, each with a `Person` label and a `name` property.
```

> The one representing Charlie Sheen also has the label `Actor` while the one representing Oliver Stone receives the label `Director`.

Reviewer 2: ```suggestion
The node representing Charlie Sheen also has the label `Actor` while the node representing Oliver Stone is assigned the label `Director`.
```

.Result
[role="queryresult",options="footer",cols="1*<m"]
|===
1+|(empty result)
1+d|Rows: 0 +
Nodes created: 2 +
Properties set: 2 +
Labels added: 4

> |===

Reviewer 2: I think this can be removed.

[[create-relationships]]
== Syntax for relationships

> You can create relationships very similarly to nodes.

Reviewer 2: ```suggestion
Relationships can also be created using the `CREATE` clause.
```

> However, relationships always need exactly one relationship type and a direction.

Reviewer 2: ```suggestion
Unlike nodes, relationships always need exactly one relationship type and a direction.
```

Beyond this, they may also have properties and names.

> They may not be quantified or carry `WHERE`-subclause.

Reviewer 2: Again, I am unsure about the saying relationships have "names"? Maybe I am missing something?

Also, maybe we should elaborate on what quantified means in this context? I am guessing you mean you need to create them one by one?

```suggestion
Similar to nodes, relationships can be assigned properties and names.
They may not be quantified or carry a `WHERE`-subclause.
```

Author: I removed the sentence on quantified relationships and `WHERE`-clauses as it only made sense when defining a pattern to follow a `CREATE`.

.Query
[source, cypher, indent=0]
----
CREATE (charlie:Person:Actor {name: 'Charlie Sheen'})-[:ACTED_IN {role: 'Bud Fox'}]->(wallStreet:Movie {title: 'Wall Street'})<-[:DIRECTED]-(oliver:Person:Director {name: 'Oliver Stone'})
----

> Creates not only the nodes for Charlie Sheen and Oliver Stone as well as the one for the movie Wall Street but also the relationships of type `ACTED_IN` and `DIRECTED` between them.

Reviewer 2: ```suggestion
This query creates the `Person` nodes for Charlie Sheen and Oliver Stone and the `Movie` node for Wall Street.
It also created the relationships of the types `ACTED_IN` and `DIRECTED` between them.
```

.Result
[role="queryresult",options="footer",cols="1*<m"]
|===
1+|(empty result)
1+d|Rows: 0 +
Nodes created: 3 +
Relationships created: 2 +
Properties set: 5 +

> Labels added: 5

Reviewer 2: Same as before, I think this could be removed.

|===

[[reusing-variables]]
== Reusing variables
The previous example created a whole path at once.

> Note, that these newly created elements are now disconnected from the rest of the graph.

Reviewer 2: I am unsure about what you mean with the word "now" in the second sentence?
Do you mean that these newly created nodes and relationships are not connected to what was previously in the graph?

```suggestion
The previous example created a path between the specified nodes.
Note, that these newly created nodes and relationships are not connected to what was previously in the graph.
```

Author: Yes, that captures it. "Now" was meant like "after the `CREATE` took effect".

> To instead connect them to already existing data, you can refer to variables from other parts of the query.

Reviewer 2: Is this a correct interpretation?

```suggestion
To connect them to already existing data, bind the desired nodes and relationships to variables.
These variables can then be passed along to subsequent clauses in a query that target pre-existing elements in the graph.
```

.Query
[source, cypher, indent=0]
----
MATCH (charlie:Person {name: 'Charlie Sheen'}), (oliver:Person {name: 'Oliver Stone'})
CREATE (charlie)-[:ACTED_IN {role: 'Bud Fox'}]->(wallStreet:Movie {title: 'Wall Street'})<-[:DIRECTED]-(oliver)
----

> First looks up the two nodes representing Charlie Sheen and Oliver Stone in order to then connect them in the same way as we have done in the previous query.

Reviewer 2: ```suggestion
In this example, the `MATCH` clause finds the nodes Charlie Sheen and Oliver Stone and binds them to the `charlie` and `oliver` variables respectively. 
These variables are then passed along to the subsequent `CREATE` clause, which creates new relationships from the bound nodes.
```

You can also reuse variables from the same `CREATE`, both in the same or a later clause.
This way, you can, for example, define constructs that are more complex than just a linear path.

.Query
[source, cypher, indent=0]
----
CREATE p = (charlie:Person:Actor {name: 'Charlie Sheen'})-[:ACTED_IN {role: 'Bud Fox'}]->(wallStreet:Movie {title: 'Wall Street'})<-[:DIRECTED]-(oliver:Person:Director {name: 'Oliver Stone'}), (wallStreet)<-[:ACTED_IN {role: 'Gordon Gekko'}]-(michael:Person:Actor {name: 'Michael Douglas'})
RETURN length(p)
----

Creates all three nodes for Charlie Sheen, Oliver Stone and Michael Douglas and connects them all to the node representing the Wall Street movie.
It then returns the length of the path from Charlie Sheen to Oliver Stone.

> Note that when repeating a node's name, you may not apply labels or properties to the repetition.

Reviewer 2: Now I am a little bit more confused about what a node's name is? In this context it seems to suggest it is not a label, but a variable?

```suggestion
Note that when repeating a node's name, you may not add labels or properties to the repetition.
```

.Query
[source, cypher, role=test-fail]
----
MATCH (charlie:Person {name: 'Charlie Sheen'})
CREATE (charlie:Actor)
----

> This query will fail because the node `charlie` already exists, and we therefore cannot `CREATE` it.

Reviewer 2: I think this makes it clearer:
```suggestion
This query will fail because the variable `charlie` has already been bound to a pre-existing node, and therefore it cannot be reused to create a new node.
```

If you intend to add a label, use the xref:clauses/set.adoc#set-set-a-label-on-a-node[`SET` clause] instead.

[[reusing-variables-properties]]
== Reusing variables in properties
The value that can be assigned to a node's or a relationship's property can be defined by an xref::syntax/expressions.adoc[expression].

////
.Parameters
[source,javascript, role=test-setup]
----
{
  "age": 23
}
----
////

.Query
[source, cypher, indent=0]
----
MATCH (person:Person)
  WHERE person.name IS NOT NULL
CREATE (anotherPerson:Person {name: person.name, age: $age})
----

> Creates a person with the same name as another person and the age from a xref:syntax/parameters.adoc[parameter] called `age`.

Reviewer 2: ```suggestion
This example created a `Person` node with the same name as another person and the age from a xref:syntax/parameters.adoc[parameter] called `age`.
```

Author: I don't mind but I tried to copy the style used in the `MATCH` page, because I thought that was the current best practice.

There, we for example have:

> By specifying a pattern with a single node and no labels, all nodes in the graph will be returned.
> Query
> 
> MATCH (n)
> RETURN n
> 
> Returns all the nodes in the database.

Now, I am just a bit unsure whether that is the preferred style.

Reviewer 2: I see what you mean, and re-reading I think both works.

Such an expression may not contain a reference to a variable that is defined in the same `CREATE` statement.

> This is defined this way so that it is always clear what value a property should have.

Reviewer 2: ```suggestion
This is to ensure that the value of a property is always clear.
```

.Query
[source, cypher, role=test-fail]
----

> CREATE (charlie {score: oliver.score + 1}), (oliver {score: charlie.score + 1})

Reviewer 2: ```suggestion
MATCH (charlie:Person {name: 'Charlie Sheen), (oliver:Person {name: 'Oliver Stone'})
CREATE (charlie {score: oliver.score + 1}), (oliver {score: charlie.score + 1})
```

Author: No, that would fail for different reasons.

----

> Tries to create nodes such that Charlie's score is higher than Oliver's and vice versa, which is a contradiction.

Reviewer 2: ```suggestion
This query tries to create nodes such that Charlie's score is higher than Oliver's and vice versa, which is a contradiction.
The query therefore fails.
```

> Therefore, queries like this are rejected.

Reviewer 2: ```suggestion
```
